bindings/go/scip/symbol.go

```diff
@@ -21,253 +20,55 @@ func IsLocalSymbol(symbol string) bool {
-		end = len(s.Symbol)
-	}
-	return nil, errors.Newf("unrecognized descriptor %q", string(s.Symbol[start:end]))
-}
-
-func (s *symbolParser) acceptIdentifier(what string) (string, error) {
-	if s.current() == '`' {
-		s.index++
-		return s.acceptBacktickEscapedIdentifier(what)
-	}
-	start := s.index
-	for s.index < len(s.Symbol) && isSimpleIdentifierCharacter(s.current()) {
-		s.index++
-	}
-	if start == s.index {
-		return "", s.error("empty identifier")
-	}
-	return string(s.Symbol[start:s.index]), nil
+// ParseSymbolUTF8 parses an SCIP string into the Symbol message.
+//
+// Unlike ParseSymbol, this skips UTF-8 validation. To customize
+// parsing behavior, use ParseSymbolUTF8With.
+func ParseSymbolUTF8(symbol beaut.UTF8String) (*Symbol, error) {
+	s := Symbol{}
+	err := ParseSymbolUTF8With(symbol, ParseSymbolOptions{IncludeDescriptors: true, RecordOutput: &s})
+	return &s, err
 }
 
-func (s *symbolParser) acceptSpaceEscapedIdentifier(what string) (string, error) {
-	return s.acceptEscapedIdentifier(what, ' ')
-}
+type emptySymbolError struct{}
 
-func (s *symbolParser) acceptBacktickEscapedIdentifier(what string) (string, error) {
-	return s.acceptEscapedIdentifier(what, '`')
+func (emptySymbolError) Error() string {
+	return "empty symbol"
 }
 
-func (s *symbolParser) acceptEscapedIdentifier(what string, escapeCharacter rune) (string, error) {
-	builder := strings.Builder{}
-	for s.index < len(s.Symbol) {
-		ch := s.current()
-		if ch == escapeCharacter {
-			s.index++
-			if s.index >= len(s.Symbol) {
-				break
-			}
-			if s.current() == escapeCharacter {
-				// Escaped space character.
-				builder.WriteRune(ch)
-			} else {
-				return builder.String(), nil
-			}
-		} else {
-			builder.WriteRune(ch)
-		}
-		s.index++
-	}
-	return "", s.error(fmt.Sprintf("reached end of symbol while parsing <%s>, expected a '%v' character", what, string(escapeCharacter)))
+type ParseSymbolOptions struct {
+	// IncludeDescriptors indicates whether the the descriptors in the symbol should also be
+	// parsed. This setting is only respected for global symbols.
+	IncludeDescriptors bool
+	// RecordOutput, if provided, is used to record the output of the symbol parser.
+	//
+	// Otherwise, only validation is performed without any extraneous heap
+	// allocations (except for optimization misses by the Go compiler).
+	RecordOutput *Symbol
 }
 
-func (s *symbolParser) acceptCharacter(r rune, what string) error {
-	if s.current() == r {
-		s.index++
-		return nil
+func ParseSymbolUTF8With(symbol beaut.UTF8String, options ParseSymbolOptions) error {
+	if symbol.Len() == 0 {
+		return emptySymbolError{}
 	}
-	return s.error(fmt.Sprintf("expected '%v', obtained '%v', while parsing %v", string(r), string(s.current()), what))
+	return parseSymbolV2(symbol, options.IncludeDescriptors, options.RecordOutput)
 }
 
-func (x *Package) ID() string {
-	return fmt.Sprintf("%s %s %s", x.Manager, x.Name, x.Version)
+func ValidateSymbolUTF8(symbol beaut.UTF8String) error {
```

Comment:
Adding this helper function for better readability at call-sites since upload ingestion requires filtering out occurrences/SymbolInformation values with malformed symbols. There is a noticeable perf difference between "always parsing" vs "validation-only", so it makes sense to also support validation-only.